### PR TITLE
[addons] add new IAddonInstanceHandler

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description
This class becomes in future used as parent on the various binary
addon instances as alternative to the CAddonDll who can no more
used as parent.

Currently this class include only the minimum but predefine the
coming multiple instance support.

Also is screensaver changed to new way and nearly in the style on end.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by AlwinEsch** [comment]
Needed changes for this request are done. @Paxxi @Montellese @ksooo is this style OK?

File: xbmc/addons/AddonInstanceHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Kodi; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "AddonInstanceHandler.h"
+
+#include "utils/log.h"
```
Comment (ksooo) [inline xbmc/addons/AddonInstanceHandler.cpp]: No functionality from utils/logs.h used in this ccp file, as far as I can see. Could be removed, imo.

File: xbmc/addons/AddonInstanceHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+IAddonInstanceHandler::~IAddonInstanceHandler()
+{
+}
+
+bool IAddonInstanceHandler::CreateInstance(KODI_HANDLE instance)
+{
+  if (m_addon && m_addon->CreateInstance(m_type, m_instanceId, instance, m_parentInstance) == ADDON_STATUS_OK)
+    return true;
+  return false;
```
Comment (ksooo) [inline xbmc/addons/AddonInstanceHandler.cpp]: Why not just <code>return (m_addon && m_addon->CreateInstance(m_type, m_instanceId, instance, m_parentInstance) == ADDON_STATUS_OK)</code>?

File: xbmc/addons/AddonInstanceHandler.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    const ADDON_TYPE UsedType() const { return m_type; }
+    const std::string& InstanceID() { return m_instanceId; }
+
+    std::string ID() const { return m_addon->ID(); }
+    std::string Name() const { return m_addon->Name(); }
+    std::string Author() const { return m_addon->Author(); }
+    std::string Icon() const { return m_addon->Icon(); }
+    std::string Path() const { return m_addon->Path(); }
+    std::string Profile() const { return m_addon->Profile(); }
+    AddonVersion Version() const { return m_addon->Version(); }
```
Comment (ksooo) [inline xbmc/addons/AddonInstanceHandler.h]: In the cpp file you are checking m_addon for nullptr before accessing it. Why not in the header file?

**Comment by AlwinEsch** [comment]
@ksooo have added a check to the functions on header, further is a new commit there who translate the addon status to a string and becomes used on IAddonInstanceHandler to bring a error.

The `m_addon` as nullptr brings now a LOGFATAL, normally should this never happen.

**Comment by ksooo** [comment]
@AlwinEsch From my point of view okay to merge now. cannot really judge on the screensaver changes but I guess this is good to go as well.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
jenkins build this please

**Pull request merged**